### PR TITLE
Solutions to assignment 1

### PR DESCRIPTION
**Pull request by dianatanase**

### CODE REVIEW

File: dianatanase/assignment1/Q1.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	// To tackle the insensitive case, transform words in all lowerCase characters and use the case sensitive method.
+	private static boolean areWordsAnagramsNotSensitive(String word1, String word2){
+		return areWordsAnagramsSensitive(word1.toLowerCase(), word2.toLowerCase());
+	}
+	
+	private static boolean areWordsAnagramsSensitive(String word1, String word2){
+		if(word1.length() != word2.length()){
+			return false;
+		}
+		else{
```
Comment (Thomas-de-Rivaz) [inline dianatanase/assignment1/Q1.java]: No need to have the else{} block; as the previous check would return already. Less indentation is better.

File: dianatanase/assignment1/Q1.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+			
+			for(int i = 0; i < size_word1; i++){
+				char letter = word1.charAt(i);
+				//If the letter exists in map, update it.
+				if(map_word1.containsKey(letter)){
+					int freq = map_word1.get(letter);
+					freq++;
+					map_word1.replace(letter, freq);
+				}
+				//If the letter does not exist in map, put it.
```
Comment (Thomas-de-Rivaz) [inline dianatanase/assignment1/Q1.java]: It is neater to write
`} else { // If the letter does not exist in map, put it.`
because then the comment is part of the next block.

File: dianatanase/assignment1/Q1.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+					map_word1.replace(letter, freq);
+				}
+				// Letter in word2 does not appear in word1
+				else{
+					return false;
+				}
+			}
+			
+			// Check there is no letter in word1 that appears more times than in word2 
+			for(Character letter: map_word1.keySet()){
```
Comment (Thomas-de-Rivaz) [inline dianatanase/assignment1/Q1.java]: If the words are equal length, is it even possible for this block to happen? As the code should have quit in the previous block when searching for letters.

File: dianatanase/assignment1/Q1.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+					}
+				}	
+			}
+			// No anagram found in sent1 for word2
+			if(!found)
+				return false;
+		}
+		
+		// Check there is no word left in sent1 that was not an anagram of any word in sent2
+		for(String word: map_sent1.keySet()){
```
Comment (Thomas-de-Rivaz) [inline dianatanase/assignment1/Q1.java]: Again, I do not believe that this block is possible to enter.

File: dianatanase/assignment1/Q2.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+	SimpleLinkedList(){}
+	
+	SimpleLinkedList(T rootData){
+		Node<T> rootNode = new Node<T>(rootData, null);
+		root = rootNode;
+		last = root;
+		size = 1;
+	}
+	
+	public int getSize(){
```
Comment (Thomas-de-Rivaz) [inline dianatanase/assignment1/Q2.java:48]: How would you change your code if the Linked List provided didn't contain a pre-calculated size method?

File: dianatanase/assignment1/Q2.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+		return curNode.getData();
+	}
+}
+
+public class Q2<T>{
+	
+	// Compute the size of the list
+	// Calculate how many elements to traverse from left to the kth elem required.
+	// Give as argument the list (with Node objects) and the query k.
+	public static <T> T getElem(int k, SimpleLinkedList<T> list){
```
Comment (Thomas-de-Rivaz) [inline dianatanase/assignment1/Q2.java]: the name getElem is misleading. As it is returning the kTh from last element. Maybe have a more descriptive name.

**Review by Thomas-de-Rivaz — APPROVED** [review]
Very good solutions. Think about adding tests in your next assignment that demonstrate the edge cases.
